draft-ietf-tls-esni.md

```diff
@@ -306,6 +310,118 @@ servers to rotate the keys often and improve forward secrecy.
+                opaque ipv6Address[16];
+            }
+        }
+    } Address;
+
+    struct {
+        HostName name;
+        Address address_set<0..2^16-1>;
+    } HostPointer;
+~~~~
+
+name
+: The terminal name for which resolution must yield valid A or
+AAAA records pointing to a host which holds the private ESNI key.
+
+address_set
+: An optional list of Address entries, each containing a single IPv4 or IPv6
+CIDR block, possibly of length 1.
+
+Use of this structure during the ESNI resolution algorithm is described in {{esni-resolution}}.
+
+### Resolution Algorithm
+
+Clients obtain ESNI records by querying the DNS for ESNI-enabled server domains.
+In cases where the domain of the A or AAAA records being resolved do not match the
+SNI Server Name, such as when {{!RFC7838}} is being used, the alternate domain should
+be used for querying the ESNI TXT record.
+
+Clients SHOULD initiate ESNI queries in parallel alongside normal A or AAAA queries.
+The following algorithm describes a procedure by which clients can process ESNIKeys
+responses as they arrive to produce addresses for ESNI-capable hosts.
+
+~~~
+1. If an ESNIKeys response arrives before an A or AAAA response, and the ESNIKeys
+response contains a "host_pointer" extension with a full address, clients SHOULD 
+initiate TLS with ESNI to the provided address(es).
+
+2. If an A or AAAA response arrives before the ESNIKeys response, clients SHOULD
+wait up to CD seconds before initiating TLS to either address. (Clients may begin
+TCP connections in this time. QUIC connections should wait.) If an ESNIKeys
+response does not arrive in this time, clients SHOULD initiate TLS without ESNI
+to the provided address(es).
+
+3. If an ESNIKeys response with a "host_pointer" extension arrives before this time, 
+and if the address netmask matches that in the corresponding A or AAAA response, 
+then clients SHOULD initiate TLS with ESNI to said address(es). (An ESNIKeys record 
+with a full address has a full netmask, and therefore the match must be exact.) 
+Clients MAY delay this check until both A and AAAA responses arrive, or if a 
+Resolution Delay timeout occurs, according to {{RFC8305}}.
+
+4. If the address netmask does not match the corresponding A or AAAA response,
+and the A or AAAA response yielded a terminal CNAME that matches host_pointer.name,
+then clients SHOULD initiate TLS using ESNI to the address(es) in the A or AAAA records.
+
+5. If the address netmask does not match the corresponding A or AAAA response,
+and neither the A or AAAA queries yielded a CNAME that matches host_pointer.name,
+clients SHOULD resolve the address of host_pointer.name using an
+address type that matches the A or AAAA response. Clients SHOULD initiate TLS
+using ENSI to the resulting address(es) when A or AAAA resolution completes.
+
+6. If the "host_pointer" address set is empty, clients SHOULD resolve the
+address(es) of host_pointer.name and initiate TLS with ESNI to the resulting 
+address(es).
```

Comment:
Isn't this just a subcase of the previous points?  If the address_set is empty, then by definition there's no overlap with the A/AAAA results.  So either the name occurs in the CNAME chain (4) or you resolve the name (5).

Reply:
Yes, technically. I left it as is to be explicit. I can remove it if we think it's excessive.

Reply:
Maybe instead of listing it as a separate step, just turn it into an unnumbered comment that an empty address set will always land in (4) or (5).